src/query_plan/operation.rs

```diff
@@ -897,6 +1002,50 @@ pub(crate) mod normalized_fragment_spread_selection {
         pub(crate) selection_id: SelectionId,
     }
 
+    impl NormalizedFragmentSpreadData {
+        pub(crate) fn type_if_added_to(
+            &self,
+            parent_type: &CompositeTypeDefinitionPosition,
+            schema: &ValidFederationSchema,
+        ) -> Result<Option<CompositeTypeDefinitionPosition>, FederationError> {
```

Comment:
Remaining question: I'm not sure how to port this from JS. Inheritance made this hard to parse out. A pointer on where to look would be helpful.

Reply:
You should add those functions to top level impls (outside of inner mods)

* `NormalizedFieldData` (which would be invoked from field selection) should implement `type_if_added_to` -> [JS field `typeIfAddedTo`](https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L349-L362)
* On JS side of things `FragmentElement` is used to represent both inline fragment data and spread data so it should be the same logic for both.... but since `can_add_to` is always true for spreads, we just need to implement `casted_type_if_added_to` for `NormalizedInlineFragmentData` (as it would be invoked from its `can_add_to`) -> [JS inline fragment `castedTypeIfAddedTo`](https://github.com/apollographql/federation/blob/main/internals-js/src/operations.ts#L524-L531)